File: parity_test.py
# -*- coding: utf-8 -*-
"""parity_test.py — Python(oasis_core) と JS(model.js) が同じ予測を出すか検証する。

    python parity_test.py          # node が要る

特徴量を1つ足すたびに、Python 側だけ直して JS 側を忘れる事故が起きる。
それを機械的に止めるためのテスト。乱数の係数でモデルを作り、同じ馬を
両方に食わせて予測値が一致するかを見る。
"""
import json
import os
import subprocess
import sys

import numpy as np
from sklearn.linear_model import Ridge

# パイプ経由で呼ばれると Windows では出力が cp932 になり、記号で落ちる。
try:
    sys.stdout.reconfigure(encoding='utf-8', errors='replace')
    sys.stderr.reconfigure(encoding='utf-8', errors='replace')
except (AttributeError, ValueError):
    pass

HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, HERE)
import oasis_core as oc  # noqa: E402

TOL = 1e-9


# 図鑑30種の説明文。`@` が % の数値に置き換わる。
ITEM_DESCS = [
    ('ロケットスタート', '序盤のスピードが@%上昇', 'gear_rocket_start'),
    ('中盤加速', '中盤のパワーが@%上昇', 'gear_mid_acceleration'),
    ('末脚', '終盤のパワーが@%上昇', 'gear_final_kick'),
    ('二の脚', '中盤開始後200mのスピードが@%上昇', 'gear_second_gear'),
    ('追い込み', '終盤に下位半分ならパワーが@%上昇', 'gear_late_charge'),
    ('競り合い', '20m以内にライバルがいるとパワーが@%上昇', 'gear_duel_spirit'),
    ('復讐刻印', '追い抜かれてから200mの間、パワーが@%上昇', 'gear_revenge_mark'),
    ('亡者の追走', '中盤で下位半分ならスピードが@%上昇', 'gear_grave_chase'),
    ('孤影の疾走', '50m以内に相手がいないとスピードが@%上昇', 'gear_lone_run'),
    ('血走り', '残りスタミナ25%以下でスピードが@%上昇', 'gear_blood_rush'),
    ('骨砕き', '残りスタミナ25%以下でパワーが@%上昇', 'gear_bone_break'),
    ('王殺し', '先頭から20m以内の2位以下でスピードが@%上昇', 'gear_king_slayer'),
    ('首位の呪い', '先頭の間、スピードが@%上昇するがスタミナ消費も増加', 'gear_leader_curse'),
    ('終焉加速', '残り300mでスピードが@%上昇', 'gear_end_accel'),
    ('芝啜り', '芝でスピードとパワーがそれぞれ@%上昇', 'gear_turf_gnaw'),
    ('泥啜り', 'ダートでスピードとパワーがそれぞれ@%上昇', 'gear_dirt_gnaw'),
    ('俊足の加護', 'スピードが常時@%上昇', 'charm_speed'),
    ('持久の加護', 'スタミナが常時@%上昇', 'charm_stamina'),
    ('剛力の加護', 'パワーが常時@%上昇', 'charm_power'),
    ('調和の加護', '全ステータスが常時@%上昇', 'charm_balance'),
    ('禍福の天秤', '全ステータスが常時@%上昇', 'charm_scale'),
    ('省エネの加護', 'スタミナ消費が常時@%減少', 'charm_energy_saver'),
    ('安定の加護', 'レース中の乱数幅を@%狭める', 'charm_consistency'),
    ('苦痛慣れ', 'スタミナ0でも減速しにくくなる', 'charm_pain_tolerance'),
    ('魂継ぎ', '残り20%以下で一度だけ最大スタミナの@%を回復', 'charm_soul_relay'),
    ('黄昏の護り', '終盤のパワーが@%上昇', 'charm_twilight_guard'),
    ('夜明けの護り', '序盤のスタミナ評価が@%上昇', 'charm_dawn_guard'),
    ('先導祈願', '先頭の間、スピードが@%上昇', 'charm_lead_prayer'),
    ('逆境祈願', '下位半分の間、スピードとパワーがそれぞれ@%上昇', 'charm_adversity'),
    ('禍福転倒', '残りスタミナ25%以下で全ステータスが@%上昇', 'charm_reversal'),
]



def main():
    spec = oc.load_passive_spec(os.path.join(HERE, 'passive_spec.json'))
    names = oc.feature_names(spec)
    rng = np.random.default_rng(0)
    # 適当な係数のモデル（学習済みログが無くても検証できるようにする）
    X = rng.normal(size=(400, len(names)))
    m = Ridge(alpha=1.0).fit(X, rng.normal(size=400))
    m.coef_ = rng.normal(scale=0.3, size=len(names))
    bundle = {'ok': True, 'model': m, 'feature_names': names, 'spec': spec,
              'race_sigma': 0.2, 'tri_sigma': 0.2, 'n_races': 0}
    payload = oc.export_model_json(bundle, os.path.join(HERE, '_parity_model.json'))

    pas = [p for p in oc.PASSIVE_NAMES]
    cases = []
    for i in range(60):
        dist = oc.DIST_LIST[i % 4]
        track = oc.TRACK_LIST[i % 2]
        horses = []
        for j in range(6):
            k = rng.integers(0, len(pas), size=2)
            horses.append({
                'name': f'h{j}', 'species': f'sp{j}',
                'speed': int(rng.integers(30, 160)), 'power': int(rng.integers(30, 160)),
                'stamina': int(rng.integers(20, 120)),
                'condition': ['好調', '普通', '不調'][int(rng.integers(0, 3))],
                'passives': [pas[k[0]], pas[k[1]]],
            })
        base = oc.predict_base(bundle, [dict(h, passives=tuple(h['passives']))
                                        for h in horses], dist, track)
        cases.append({'dist': dist, 'track': track, 'horses': horses,
                      'base': [float(x) for x in base]})
    # --- 装備・お守りの効果（図鑑30種）---
    # ⚠ カタログに無い効果名は Python が説明文から推測するのに対し、JS は**乗せない**。
    #    意図的な差（推測は race 2097 で15倍の誤りを出した実績がある）なので、
    #    検証はカタログにある30種だけを対象にする。
    # 馬場限定（芝啜り／泥啜り）は ctx の有無で答えが変わるので3通りとも回す。
    # 先頭でだけ効く装備は ctx の 1着確率で duty が変わるので、強い馬・弱い馬の両方を回す。
    CTXS = [None, {'dist': '短距離', 'track': '芝'}, {'dist': 'マイル', 'track': 'ダート'},
            {'dist': '長距離', 'track': '芝', 'p_win': 0.85, 'n_field': 13},
            {'dist': '短距離', 'track': 'ダート', 'p_win': 0.02, 'n_field': 16}]
    item_cases = []
    for i, (label, desc, key) in enumerate(ITEM_DESCS):
        pct = round(1.0 + (i % 9) * 0.9, 1)
        d = desc.replace('@', str(pct))
        for ctx in CTXS:
            got = oc.item_effect_spec(f'{label}：{d}', key, spec, ctx)
            item_cases.append({'label': label, 'desc': d, 'key': key, 'ctx': ctx,
                               'py': (None if got is None else
                                      {k: float(v) for k, v in got.items()})})
    # --- 単勝（下限オッズの判定 → 希薄化を織り込んだ配分）---
    # ランダムなだけでは踏まない枝（未投票馬・取得漏れ NaN・下限に張り付いた大本命・
    # 自分がプールの大半を持っている状態）を、モードを回して必ず通す。
    def _nn(x):
        """JSON に NaN は書けないので null にする（JS 側で NaN に戻す）。"""
        x = float(x)
        return None if x != x else x

    win_cases = []
    for i in range(48):
        mode = i % 4
        a = np.ones(int(rng.integers(6, 13 if mode == 2 else 15))) * 1.2
        if mode == 1:
            a[0] = 20.0                                   # 1頭にシェアが集中＝下限に張り付く大本命
        od = list(np.round(1.0 / rng.dirichlet(a), 2))
        if mode == 2:
            od += [oc.UNBET_ODDS, oc.UNBET_ODDS]          # 未投票の馬（Σ(1/od) は 1 のまま）
        od = np.where(np.array(od, dtype=float) < oc.ODDS_FLOOR, oc.ODDS_FLOOR, od)
        if mode == 2 and i % 12 != 2:
            # Σ(1/od) を 1 から外し、下限表示の馬を「未投票」と言い切れない枝
            #（中間帯 / 下限が複数いて大本命を特定できない）を通す。
            od = np.where(od > oc.ODDS_FLOOR, np.round(od * (1.05 if i % 12 == 6 else 1.6), 2), od)
        if mode == 3:
            od[rng.choice(len(od), size=2, replace=False)] = np.nan
        n = len(od)
        my_amounts = np.zeros(n)
        if i % 3 == 0:                                    # 自分で買った下限表示馬（判別できる側）
            my_amounts[0 if mode == 1 else int(rng.integers(0, n))] = \
                float(rng.integers(1, 10)) * oc.WIN_STAKE_UNIT
        floor = 1.0 if i % 2 else oc.ODDS_FLOOR
        mkt = oc.market_win_prob(od, floor=floor)
        diag = oc.diagnose_floor_odds(od, my_amounts)

        win_p = rng.dirichlet(np.ones(n) * 3.0)
        pool = float([oc.WIN_POOL_SEED, 50_000, 1_000_000, 20_000][mode])
        my_units = np.zeros(n, dtype=int)
        if i % 2:
            for j in rng.choice(n, size=2, replace=False):
                my_units[j] = int(rng.integers(1, 6))
        if i % 8 == 5:                                    # 自分がプールの大半を持っている状態
            pool, my_units = 20_000.0, np.zeros(n, dtype=int)
            my_units[0] = 15
        if i % 16 == 7:                                   # 既に上限まで購入済み（note を返す枝）
            my_units = np.zeros(n, dtype=int)
            my_units[0] = oc.WIN_MAX_TOTAL_UNITS
        if i % 16 == 3:                                   # プール未測定（推奨なしの枝）
            pool = 0.0
        bankroll = float(rng.integers(50_000, 2_000_000))
        kelly, edge_min = (0.25, 0.5)[i % 2], (0.15, 0.05, 0.0)[i % 3]
        risk_cap = (0.10, 0.30)[i % 2]
        names_w = [f'h{j}' for j in range(n)]
        picks, summ = oc.win_bet_picks_pool(
            names_w, win_p, diag['odds_eff'], pool, bankroll, kelly, edge_min,
            stake_unit=oc.WIN_STAKE_UNIT, total_units=oc.WIN_MAX_TOTAL_UNITS,
            max_units=oc.WIN_MAX_UNITS, risk_cap_frac=risk_cap,
            my_units=my_units, unbet=diag['unbet'])
        win_cases.append({
            'names': names_w, 'odds': [_nn(x) for x in od], 'floor': floor,
            'my_amounts': [float(x) for x in my_amounts],
            'mkt': (None if mkt is None else [float(x) for x in mkt]),
            'diag': {'unbet': [bool(b) for b in diag['unbet']],
                     'odds_eff': [_nn(x) for x in diag['odds_eff']],
                     'ambiguous': bool(diag['ambiguous']),
                     'residual': (None if diag['residual'] is None
                                  else float(diag['residual'])),
                     'messages': list(diag['messages'])},
            'win_p': [float(x) for x in win_p], 'pool': pool, 'bankroll': bankroll,
            'kelly': kelly, 'edge_min': edge_min, 'risk_cap_frac': risk_cap,
            'my_units': [int(x) for x in my_units],
            'picks': picks, 'summary': summ})

    # --- 3連単（1組の最適口数 → 成立組の配分 → 未成立組の1口買い）---
    # プールが初期金ちょうど／少しだけ上、inf/nan のオッズ、誰も条件を満たさない
    # edge_min など、実戦で踏む枝をモードで一巡させる。
    def _od(x):
        """inf/nan は JSON に書けないので印にする（JS 側で戻す）。"""
        x = float(x)
        if x != x:
            return None
        return 'inf' if x == float('inf') else x

    tri_cases = []
    for i in range(40):
        n_c = int(rng.integers(5, 41))
        pool = float([oc.TRIFECTA_POOL_SEED, oc.TRIFECTA_POOL_SEED + oc.STAKE_UNIT,
                      3_000_000, 420_000, 0][i % 5])
        od = np.exp(rng.uniform(np.log(1.5), np.log(5000.0), size=n_c))
        if i % 4 == 1:                                    # 弾かれるべきオッズ
            for j, bad in zip(rng.choice(n_c, size=3, replace=False),
                              (np.inf, np.nan, 1.0)):
                od[j] = bad
        safe = np.where(np.isfinite(od) & (od > 1.0), od, 100.0)
        # 1/od の 0.3〜3.0 倍＝プラスEVもマイナスEVも混ざる確率
        p = np.clip(rng.uniform(0.3, 3.0, size=n_c) / safe, 1e-6, 0.999)
        if i % 5 == 3:
            p[0] = 0.0                                    # 0<p<1 の枝
        # 添字を回して決めると周期が噛み合ってしまい（資金が小さいケースは必ず
        # 総口数1口、など）、片方の上限だけが効く枝を素通りする。乱数で振る。
        # kelly 0.03 はケリー口数が 0〜1 口に落ちる値（max(1, k_kelly) の枝）。
        bankroll = float(rng.integers(12_000, 5_000_000))
        kelly = float(rng.choice([0.03, 0.25, 0.5, 1.0]))
        max_risk = float(rng.choice([0.05, 0.10, 0.30]))
        edge_min = 3.0 if i % 4 == 3 else float(rng.choice([0.0, 0.05, 0.30]))
        budget = int(rng.choice([oc.MAX_TOTAL_UNITS, 5, 1]))
        max_per = [oc.MAX_UNITS, 3, None][int(rng.integers(0, 3))]
        cands = [(f'c{j}', float(p[j]), float(od[j])) for j in range(n_c)]
        opt = [list(oc.optimal_units_ev(float(p[j]), float(od[j]), pool,
                                        oc.STAKE_UNIT, oc.MAX_UNITS))
               for j in range(n_c)]
        alloc = oc.allocate_units_stable(
            cands, pool, bankroll, kelly, max_risk, edge_min,
            budget=budget, stake_unit=oc.STAKE_UNIT, max_per_combo=max_per)

        # 未成立組（od_of が None を返す組）だけに1口ずつ
        n_h = 8
        disp = [f'h{j}' for j in range(n_h)]
        combo_prob = {}
        for _ in range(int(rng.integers(3, 20))):
            idx = tuple(int(x) for x in rng.choice(n_h, size=3, replace=False))
            combo_prob[idx] = float(rng.uniform(0.001, 0.4))
        od_map = {tuple(disp[j] for j in idx): float(rng.uniform(2, 500))
                  for idx in list(combo_prob)[::2]}     # 半分は成立済み＝除外される
        sl = {'pMin': float(rng.choice([0.01, 0.05, 0.2])),
              'edgeMin': float(rng.choice([0.0, 0.3, 5.0])),   # 5.0 は誰も通らない
              'maxUnits': int(rng.choice([5, 0, 40])),         # 0 は即 [] を返す枝
              'remainingBudget': int(rng.choice([budget, 2, -1, 3])),  # -1 も同じ
              'pScale': float(rng.choice([1.0, 0.6]))}
        sleeve = oc.unformed_sleeve_picks(
            combo_prob, disp, lambda nm: od_map.get(nm), pool,
            p_min=sl['pMin'], edge_min=sl['edgeMin'], max_units=sl['maxUnits'],
            remaining_budget=sl['remainingBudget'],
            stake_unit=oc.STAKE_UNIT, p_scale=sl['pScale'])

        tri_cases.append({
            'pool': pool, 'bankroll': bankroll, 'kelly': kelly,
            'max_risk_frac': max_risk, 'edge_min': edge_min,
            'budget': budget, 'max_per_combo': max_per,
            'cands': [{'key': c, 'p': pp, 'od': _od(o)} for c, pp, o in cands],
            'opt': [[float(k), float(ev), _od(o)] for k, ev, o in opt],
            'alloc': {c: [float(k), float(ev), float(e)]
                      for c, (k, ev, e) in alloc.items()},
            'combo_prob': [[list(idx), pp] for idx, pp in combo_prob.items()],
            'disp': disp,
            'od_map': [['|'.join(nm), o] for nm, o in od_map.items()],
            'sleeve_opts': sl,
            'sleeve': [[list(nm), float(pp), float(e), int(k)]
                       for nm, pp, e, k in sleeve]})

    with open(os.path.join(HERE, '_parity_cases.json'), 'w', encoding='utf-8') as f:
        json.dump({'races': cases, 'items': item_cases, 'win': win_cases,
                   'tri': tri_cases}, f, ensure_ascii=False)

    try:
        # Windows の既定は cp932 で、✅ や ⚠ を書けずに落ちる。明示的に UTF-8 で受ける。
        r = subprocess.run(['node', os.path.join(HERE, 'parity_test.js')],
                           capture_output=True, text=True,
                           encoding='utf-8', errors='replace')
        code, out, err = r.returncode, r.stdout, r.stderr
    except FileNotFoundError:
        # node が無い環境（Windows の素の状態など）。検証はできないが失敗ではない。
        code, out, err = 0, ('⚠ node が見つからないので Python↔JS の一致検証を飛ばしました。\n'
                             '   model.js を触ったなら https://nodejs.org から入れて再実行してください。\n'), ''
    sys.stdout.write(out)
    if err:
        sys.stderr.write(err)
    for f in ('_parity_model.json', '_parity_cases.json'):
        try:
            os.remove(os.path.join(HERE, f))
        except OSError:
            pass
    print(f'  特徴量 {len(names)}列 / スタミナ収支 '
          f'{"あり" if "スタミナ余り" in names else "★なし★"}')
    return code


if __name__ == '__main__':
    sys.exit(main())

File: parity_test.js
// parity_test.js — parity_test.py から呼ばれる。単体では動かない。
const fs = require('fs');
const path = require('path');
const HERE = __dirname;
const M = JSON.parse(fs.readFileSync(path.join(HERE, '_parity_model.json'), 'utf8'));
const _all = JSON.parse(fs.readFileSync(path.join(HERE, '_parity_cases.json'), 'utf8'));
const cases = _all.races, items = _all.items || [];
const OM = require(path.join(HERE, 'bookmarklets/src/model.js'));

let worst = 0, bad = 0;
for (const c of cases) {
  const js = OM.predictBase(c.horses, c.dist, c.track, M);
  for (let i = 0; i < js.length; i++) {
    const d = Math.abs(js[i] - c.base[i]);
    if (d > worst) worst = d;
    if (d > 1e-9) bad++;
  }
}
// --- 装備・お守りの効果（図鑑30種）---
let ibad = 0, iworst = 0;
for (const c of items) {
  const js = OM.itemMult({ effect_label: c.label, effect_description: c.desc, effect_key: c.key }, M, []);
  const py = c.py;
  if ((js == null) !== (py == null)) { ibad++; console.log(`  ❌ ${c.label}: JS=${JSON.stringify(js)} / PY=${JSON.stringify(py)}`); continue; }
  if (js == null) continue;
  const keys = new Set([...Object.keys(js), ...Object.keys(py)]);
  for (const k of keys) {
    const d = Math.abs((js[k] == null ? 1 : js[k]) - (py[k] == null ? 1 : py[k]));
    if (d > iworst) iworst = d;
    if (d > 1e-9) { ibad++; console.log(`  ❌ ${c.label}.${k}: JS=${js[k]} / PY=${py[k]}`); }
  }
}
console.log(`装備効果 一致検証: ${items.length}種  最大誤差 ${iworst.toExponential(2)}`);
if (ibad) { console.log(`❌ 装備効果 不一致 ${ibad}件 — model.js の itemMult が oasis_core に追随していません`); process.exit(1); }

// --- 単勝（下限オッズの判定・希薄化を織り込んだ配分）---
const wins = _all.win || [];
const _nan = v => (v === null ? NaN : v);   // Python の NaN は JSON に書けないので null で来る
let wbad = 0, wworst = 0;
const chk = (a, b, what) => {               // NaN 同士は一致、片方だけ NaN は不一致
  a = _nan(a); b = _nan(b);
  const d = (Number.isNaN(a) || Number.isNaN(b))
    ? ((Number.isNaN(a) && Number.isNaN(b)) ? 0 : Infinity) : Math.abs(a - b);
  if (d > wworst) wworst = d;
  if (d > 1e-9) { wbad++; console.log(`  ❌ ${what}: JS=${a} / PY=${b}`); }
};
const eq = (a, b, what) => {
  if (a !== b) { wbad++; console.log(`  ❌ ${what}: JS=${JSON.stringify(a)} / PY=${JSON.stringify(b)}`); }
};
for (let ci = 0; ci < wins.length; ci++) {
  const c = wins[ci], tag = `win[${ci}]`;
  const od = c.odds.map(_nan);
  const mkt = OM.marketWinProb(od, c.floor);
  eq(mkt === null, c.mkt === null, `${tag}.mkt null`);
  if (mkt && c.mkt) for (let i = 0; i < mkt.length; i++) chk(mkt[i], c.mkt[i], `${tag}.mkt[${i}]`);

  const dg = OM.diagnoseOddsFloor(od, c.my_amounts, M);
  eq(dg.ambiguous, c.diag.ambiguous, `${tag}.ambiguous`);
  chk(dg.residual, c.diag.residual, `${tag}.residual`);
  eq(JSON.stringify(dg.unbet), JSON.stringify(c.diag.unbet), `${tag}.unbet`);
  eq(JSON.stringify(dg.messages), JSON.stringify(c.diag.messages), `${tag}.messages`);
  for (let i = 0; i < od.length; i++) chk(dg.odds_eff[i], c.diag.odds_eff[i], `${tag}.odds_eff[${i}]`);

  const r = OM.winBetPicksPool(c.names, c.win_p, dg.odds_eff, c.pool, c.bankroll,
    c.kelly, c.edge_min, { stakeUnit: M.win_stake_unit, totalUnits: M.win_max_total_units,
      maxUnits: M.win_max_units, riskCapFrac: c.risk_cap_frac,
      myUnits: c.my_units, unbet: dg.unbet });
  const picks = r[0], summ = r[1];
  eq(picks.length, c.picks.length, `${tag}.picks 件数`);
  for (let i = 0; i < Math.min(picks.length, c.picks.length); i++) {
    const a = picks[i], b = c.picks[i];
    eq(a.name, b.name, `${tag}.picks[${i}].name`);
    eq(a.unbet, b.unbet, `${tag}.picks[${i}].unbet`);
    eq(a.odds === null, b.odds === null, `${tag}.picks[${i}].odds null`);
    if (a.odds !== null && b.odds !== null) chk(a.odds, b.odds, `${tag}.picks[${i}].odds`);
    for (const k of ['p', 'eff_od', 'edge', 'units', 'stake', 'ev']) chk(a[k], b[k], `${tag}.picks[${i}].${k}`);
  }
  eq(summ === null, c.summary === null, `${tag}.summary null`);
  if (summ && c.summary) for (const k of Object.keys(c.summary)) {
    if (typeof c.summary[k] === 'number') chk(summ[k], c.summary[k], `${tag}.summary.${k}`);
    else eq(summ[k], c.summary[k], `${tag}.summary.${k}`);
  }
}
console.log(`単勝配分 一致検証: ${wins.length}件  最大誤差 ${wworst.toExponential(2)}`);
if (wbad) { console.log(`❌ 単勝配分 不一致 ${wbad}件 — model.js の単勝ロジックが oasis_core に追随していません`); process.exit(1); }

// --- 3連単（1組の最適口数・成立組の配分・未成立組の1口買い）---
const tris = _all.tri || [];
const _od = v => (v === null ? NaN : (v === 'inf' ? Infinity : v));
let tbad = 0, tworst = 0;
const tchk = (a, b, what) => {              // NaN 同士は一致、片方だけ NaN は不一致
  const d = (Number.isNaN(a) || Number.isNaN(b))
    ? ((Number.isNaN(a) && Number.isNaN(b)) ? 0 : Infinity) : Math.abs(a - b);
  if (d > tworst) tworst = d;
  if (d > 1e-9) { tbad++; console.log(`  ❌ ${what}: JS=${a} / PY=${b}`); }
};
const teq = (a, b, what) => {
  if (a !== b) { tbad++; console.log(`  ❌ ${what}: JS=${JSON.stringify(a)} / PY=${JSON.stringify(b)}`); }
};
for (let ci = 0; ci < tris.length; ci++) {
  const c = tris[ci], tag = `tri[${ci}]`;
  const cands = c.cands.map(x => ({ key: x.key, p: x.p, od: _od(x.od) }));
  for (let i = 0; i < cands.length; i++) {
    const js = OM.optimalUnitsEv(cands[i].p, cands[i].od, c.pool, M.stake_unit, M.max_units);
    const py = c.opt[i];
    for (let k = 0; k < 3; k++) tchk(js[k], _od(py[k]), `${tag}.opt[${i}][${k}]`);
  }

  const alloc = OM.allocateUnitsStable(cands, c.pool, c.bankroll, c.kelly,
    c.max_risk_frac, c.edge_min,
    { budget: c.budget, stakeUnit: M.stake_unit, maxPerCombo: c.max_per_combo });
  // Python の dict は挿入順を保つ。並びが違えば貪欲法の同点処理がズレている。
  teq(Array.from(alloc.keys()).join(','), Object.keys(c.alloc).join(','), `${tag}.alloc 並び`);
  for (const k of Object.keys(c.alloc)) {
    const a = alloc.get(k);
    if (!a) { tbad++; console.log(`  ❌ ${tag}.alloc[${k}]: JS に無い`); continue; }
    for (let j = 0; j < 3; j++) tchk(a[j], c.alloc[k][j], `${tag}.alloc[${k}][${j}]`);
  }

  const odMap = new Map(c.od_map);
  const o = c.sleeve_opts;
  const sleeve = OM.unformedSleevePicks(c.combo_prob, c.disp,
    nm => (odMap.has(nm.join('|')) ? odMap.get(nm.join('|')) : null), c.pool,
    { pMin: o.pMin, edgeMin: o.edgeMin, maxUnits: o.maxUnits,
      remainingBudget: o.remainingBudget, stakeUnit: M.stake_unit, pScale: o.pScale });
  teq(sleeve.length, c.sleeve.length, `${tag}.sleeve 件数`);
  for (let i = 0; i < Math.min(sleeve.length, c.sleeve.length); i++) {
    teq(sleeve[i][0].join('|'), c.sleeve[i][0].join('|'), `${tag}.sleeve[${i}].names`);
    for (let j = 1; j < 4; j++) tchk(sleeve[i][j], c.sleeve[i][j], `${tag}.sleeve[${i}][${j}]`);
  }
}
console.log(`3連単配分 一致検証: ${tris.length}件  最大誤差 ${tworst.toExponential(2)}`);
if (tbad) { console.log(`❌ 3連単配分 不一致 ${tbad}件 — model.js の3連単ロジックが oasis_core に追随していません`); process.exit(1); }

const n = cases.reduce((s, c) => s + c.base.length, 0);
console.log(`Python↔JS 一致検証: ${n}頭  最大誤差 ${worst.toExponential(2)}`);
if (bad) { console.log(`❌ 不一致 ${bad}件 — model.js が oasis_core に追随していません`); process.exit(1); }
console.log('✅ 一致');
